diesel/src/sqlite/connection/raw.rs

```diff
@@ -194,12 +251,150 @@ extern "C" fn run_custom_function<F>(
     }
 }
 
+// Need a custom option type here, because the std lib one does not have guarantees about the discriminate values?
+#[repr(u8)]
+enum OptionalAggregator<A> {
+    // Discriminant is 0
+    #[allow(dead_code)]
+    None,
+    Some(A),
+}
+
+#[allow(warnings)]
+extern "C" fn run_aggregator_step_function<ArgsSqlType, RetSqlType, Args, Ret, A>(
+    ctx: *mut ffi::sqlite3_context,
+    num_args: libc::c_int,
+    value_ptr: *mut *mut ffi::sqlite3_value,
+)
+where
+    A: Aggregator<Args, Output=Ret> + 'static + Send,
+    Args: Queryable<ArgsSqlType, Sqlite>,
+    Ret: ToSql<RetSqlType, Sqlite>,
+    Sqlite: HasSqlType<RetSqlType>,
+{
+    static NULL_AG_CTX_ERR: &str = "An unknown error occurred. sqlite3_aggregate_context returned a null pointer. This should never happen.";
+
+    unsafe {
+        let aggregate_context = ffi::sqlite3_aggregate_context(ctx, std::mem::size_of::<OptionalAggregator<A>>() as i32);
+        let mut aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
+        let aggregator = match aggregate_context {
+            // Check with someone that does more unsafe code reviews if this is ok. We do
+            // only access the tag of an zeroed variant here, right? That tag is guaranteed to be zero for
+            // our case and this variant.
+            Some(ref mut a) => match a.as_mut() {
+                &mut OptionalAggregator::Some(ref mut agg) => agg,
+                &mut OptionalAggregator::None => {
+                    // Not sure if we should use `write_unaligned` instead
+                    ptr::write(a.as_mut(), OptionalAggregator::Some(A::default()));
+                    if let &mut OptionalAggregator::Some(ref mut agg) = a.as_mut() {
+                        agg
+                    } else {
+                        unreachable!("We've written the aggregator above to that location, it must be there")
+                    }
+                }
+            },
```

Comment:
Wouldn't we need to keep the nested matches to capture the `a` that is referenced in the the `Some(&mut OptionalAggregator::None` arm?

Reply:
A right, that's a `NonNull` in there. I think we can make this work with one match nevertheless, by doing something like:

```rust
match aggregator_context.map(|a| &mut *a.as_ptr()) {
    Some(OptionalAggregator::Some(ref mut agg)) => agg,
    Some(OptionalAggregator::None) => {…}
    None => …
}
```

Reply:
I have this right now:
```
        let aggregator = match aggregate_context.map(|a| &mut *a.as_ptr()) {
            Some(&mut OptionalAggregator::Some(ref mut agg)) => agg,
            Some(mut a_ptr @ &mut OptionalAggregator::None) => {
                ptr::write_unaligned(a_ptr as *mut _, OptionalAggregator::Some(A::default()));
                if let &mut OptionalAggregator::Some(ref mut agg) = a_ptr {
                    agg
                } else {
                    unreachable!("We've written the aggregator above to that location, it must be there")
                }
            },
            None => ...
        };
```
which passes the tests. Do you have any good ways to test memory leaks? I tried (naively) just running `valgrind cargo test --no-default-features --features "sqlite" register_aggregate_function` but that didn't really work

Reply:
You can try with [the leak sanitizer](https://github.com/rust-lang/rustc-dev-guide/blob/master/src/sanitizers.md) provided by rustc. Otherwise the unsafe code looks sound for me. I will try to get a second opinion from someone else in the reviewer team on this, so cc @diesel-rs/reviewers: It would be great if someone could have a look at the unsafe code involved here.